snapshots/wb/2024-01-17/pip_api.py

```diff
@@ -0,0 +1,1568 @@
+POV_LINES_COUNTRIES = poverty_lines_countries()
+POV_LINES_REGIONS = poverty_lines_regions()
+
+# # DEBUGGING
+# PPP_VERSIONS = [2017]
+# POV_LINES_COUNTRIES = [1, 1000, 25000, 50000]
+# POV_LINES_REGIONS = [1, 1000, 25000, 50000]
+
+
+@click.command()
+@click.option(
+    "--live-api/--internal-api",
+    default=True,
+    type=bool,
+    help="Select live (1) or internal (0) API",
+)
+# @click.option("--path-to-file", prompt=True, type=str, help="Path to local data file.")
+def run(live_api: bool) -> None:
+    if live_api:
+        wb_api = WB_API("https://api.worldbank.org/pip/v1")
+    else:
+        wb_api = WB_API("https://apiv2qa.worldbank.org/pip/v1")
+
+    # Generate percentiles by extracting the raw files and processing them afterward
+    df_percentiles = generate_consolidated_percentiles(generate_percentiles_raw(wb_api), wb_api)
+
+    # Generate relative poverty indicators file
+    df_relative = generate_relative_poverty(wb_api)
+
+    # Generate key indicators file and patch medians
+    df = generate_key_indicators(wb_api)
+    df = median_patch(df, country_or_region="country")
+
+    # Add relative poverty indicators and decile thresholds to the key indicators file
+    df = add_relative_poverty_and_decile_threholds(df, df_relative, df_percentiles)
+
+
+class WB_API:
+    def __init__(self, api_address, check_health=False):
+        self.api_address = api_address
+        self.check_health = check_health
+
+    def health_check(self):
+        return pd.read_json(f"{self.api_address}/health-check")[0][0]
+
+    def api_health(self):
+        """
+        Check if the API is running.
+        """
+        if not self.check_health:
+            return
+
+        # Initialize repeat counter
+        repeat = 0
+
+        # health comes from a json containing the status
+        health = self.health_check()
+
+        # If the status is different to "PIP API is running", repeat the request until MAX_REPEATS
+        while health != "PIP API is running" and repeat < MAX_REPEATS:
+            repeat += 1
+
+        if repeat >= MAX_REPEATS:
+            # If the status is different to "PIP API is running" after MAX_REPEATS, log fatal error
+            raise AssertionError(f"Health check: {health} (repeated {repeat} times)")
+
+    def versions(self):
+        return memory.cache(pd.read_csv)(f"{self.api_address}/versions?format=csv")
+
+    def get_table(self, table):
+        return pd.read_csv(f"{self.api_address}/aux?table={table}&long_format=false&format=csv")
+
+    def fetch_csv(self, url):
+        return _fetch_csv(f"{self.api_address}{url}")
+
+
+@retry(wait=wait_random_exponential(multiplier=1), stop=stop_after_attempt(MAX_REPEATS))
+def _get_request(url: str) -> requests.Response:
+    response = requests.get(url, timeout=TIMEOUT)
+    if response.status_code != 200:
+        log.info("fetch_csv.retry", url=url)
+        raise Exception("API timed out")
+
+    response = requests.get(url, timeout=TIMEOUT)
```

Comment:
Coderabbit was right! There's identical code, we can remove one block (it was probably me when I was working on this).